web-console/src/utils/ingestion-spec.tsx

```diff
@@ -0,0 +1,1117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Code } from '@blueprintjs/core';
+import { number } from 'prop-types';
+import * as React from 'react';
+
+import { Field } from '../components/auto-form';
+import { ExternalLink } from '../components/external-link';
+
+import { TIMESTAMP_FORMAT_VALUES } from './druid-time';
+import { deepGet, deepSet } from './object-change';
+
+export interface IngestionSpec {
+  type?: IngestionType;
+  dataSchema: DataSchema;
+  ioConfig: IoConfig;
+  tuningConfig?: TuningConfig;
+}
+
+export type IngestionType = 'kafka' | 'kinesis' | 'index_hadoop' | 'index' | 'index_parallel';
```

Comment:
It seems like it would be really useful in the future to figure out a way to populate this stuff based on the set of loaded extensions, but I'm not sure the best mechanism to do that at this point.

Reply:
Well that is just a type, so in the future it would probably become `string`. I was planning of proposing a system for extending the data loader to arbitrary extensions. So that every extension could declare a json blob (that I will propose) and then be able to 'register' itself with the data loader